src/main/java/org/broadinstitute/barclay/help/WDLWorkUnitHandler.java

```diff
@@ -0,0 +1,337 @@
+                            throw new RuntimeException(String.format(
+                                    "Generating WDL for tools with args with multiple type parameters is not supported " +
+                                            "(class %s for arg %s in %s has multiple type parameters).",
+                                    argumentClass,
+                                    argField.getName(),
+                                    argField.getDeclaringClass()));
+                        }
+
+                        nestedTypeClass = Class.forName(pType2.getRawType().getTypeName());
+                        wdlType = convertJavaTypeToWDLType(nestedTypeClass, wdlType, argField.getDeclaringClass().toString());
+                    } else {
+                        nestedTypeClass = Class.forName(genericTypes[0].getTypeName());
+                        wdlType = convertJavaTypeToWDLType(nestedTypeClass, wdlType, argField.getDeclaringClass().toString());
+                    }
+                    return wdlType;
+                } catch (ClassNotFoundException e) {
+                    throw new RuntimeException(String.format(
+                            "WDL generation can't find class %s for %s",
+                            pType2.getRawType().toString(),
+                            argField.getDeclaringClass()), e);
+                }
+            } else {
+                throw new RuntimeException(String.format(
+                        "Generic type must have a ParameterizedType (class %s for argument %s/%s)",
+                        argumentClass,
+                        argField.getName(),
+                        argField.getDeclaringClass()));
+            }
+        }
+
+        wdlType = convertJavaTypeToWDLType(argumentClass, wdlType, argField.getDeclaringClass().toString());
+
+        return wdlType;
+    }
+
+    /**
+     * Given a Java class representing the underlying field  type of an argument, and a human readable doc type,
+     * convert the docType to a WDL type.
+     *
+     * @param argumentClass the Class for the underlying field of the argument being converted
+     * @param docType a string representing the human readable type assigned by the Barclay doc system
+     * @param contextMessage a message describing the context for this argument, used in error reporting
+     * @return the docType string transformed to the corresponding WDL type
+     */
+    protected String convertJavaTypeToWDLType(final Class<?> argumentClass, final String docType, final String contextMessage) {
+        String convertedWDLType;
+        final Pair<String, String> typeConversionPair = transformToWDLType(argumentClass);
+        if (typeConversionPair != null) {
+            convertedWDLType = docType.replace(typeConversionPair.getKey(), typeConversionPair.getValue());
+        } else if (argumentClass.isEnum()) {
+            //NOTE: we could emit structs for all the ENUM types into a shared/common WDL file ?
```

Comment:
I like this idea a lot.

Reply:
I liked it at first too, but I abandon it because it would make the individual files less portable by introducing an inter-file dependency. Also, the CLP will ensure these are correct anyway, so it seems like overkill to have WDL ensure the type safety of these as well. Might be something to revisit in the future though.

Reply:
Ah, OK.